File: library/src/main/java/com/hansoolabs/and/view/IndexIndicator.java
package com.hansoolabs.and.view;

import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.TextView;

import com.hansoolabs.and.R;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by vulpes on 2017. 6. 12..
 */

public class IndexIndicator implements IndexScrollerView.IndexIndicator {

    public interface Adapter {
        int getTotalSize();
        int getCurrentIndex(int adapterPosition, int totalAdapterSize);
        boolean isVisible();
    }

    private static final long FADE_ANIMATION_DURATION = 150;
    private static final long VISIBLE_TIMEOUT = 400;

    private final ViewGroup parent;
    private final View frame;
    private final Adapter adapter;
    private final TextView indexView;
    private final TextView totalView;
    private final Handler handler;

    private volatile long lastTouchedTime;
    private volatile Timer timer;

    private Animation animation;

    public IndexIndicator(View frame, Adapter adapter) {
        this.parent = (ViewGroup) frame.getParent();
        this.frame = frame;
        this.adapter = adapter;
        this.indexView = (TextView) frame.findViewById(R.id.index);
        this.totalView = (TextView) frame.findViewById(R.id.total);
        handler = new Handler(Looper.getMainLooper());
    }

    @Override
    public void moveToPosition(float scrollProgress) {
        int height = parent.getHeight() - frame.getHeight();
        float y = scrollProgress * height;
        frame.setY(y);
        lastTouchedTime = System.currentTimeMillis();
    }

    private AlphaAnimation createAnimation(float from, float to) {
        AlphaAnimation anim = new AlphaAnimation(from, to);
        anim.setDuration(FADE_ANIMATION_DURATION);
        return anim;
    }

    @Override
    public void show() {
        lastTouchedTime = System.currentTimeMillis();
        startTimer();
        if (frame.getVisibility() != View.VISIBLE) {
            if (animation != null) {
                animation.cancel();
            }
            animation = createAnimation(frame.getAlpha(), 1.0f);
            animation.setAnimationListener(new Animation.AnimationListener() {
                @Override
                public void onAnimationStart(Animation animation) {

                }

                @Override
                public void onAnimationEnd(Animation animation) {
                    frame.setAlpha(1.0f);
                    IndexIndicator.this.animation = null;
                }

                @Override
                public void onAnimationRepeat(Animation animation) {

                }
            });
            frame.setVisibility(View.VISIBLE);
            frame.startAnimation(animation);
        }
    }

    @Override
    public void hide() {
        stopTimerIfExist();
        if (frame.getVisibility() != View.GONE) {
            if (animation != null) {
                animation.cancel();
            }
            frame.setVisibility(View.VISIBLE);
            animation = createAnimation(frame.getAlpha(), 0.0f);
            animation.setAnimationListener(new Animation.AnimationListener() {
                @Override
                public void onAnimationStart(Animation animation) {

                }

                @Override
                public void onAnimationEnd(Animation animation) {
                    frame.setVisibility(View.GONE);
                    IndexIndicator.this.animation = null;
                }

                @Override
                public void onAnimationRepeat(Animation animation) {

                }
            });
            frame.startAnimation(animation);
        }
    }

    @Override
    public boolean isShown() {
        return frame.isShown();
    }

    @Override
    public boolean isAvailable() {
        return adapter.isVisible() && adapter.getTotalSize() > 0;
    }

    @Override
    public void updateIndex(int current, int total) {
        int index = adapter.getCurrentIndex(current, total);
        int size = adapter.getTotalSize();
        indexView.setText(String.valueOf(index));
        totalView.setText(String.valueOf(size));
    }

    private synchronized void startTimer() {
        stopTimerIfExist();
        timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                if (lastTouchedTime + VISIBLE_TIMEOUT < System.currentTimeMillis()) {
                    stopTimerIfExist();
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            hide();
                        }
                    });
                }
            }
        }, 0, 100);
    }

    private synchronized void stopTimerIfExist() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }
}

File: library/src/main/kotlin/com/cherrytree/skinnyyoga/view/IndexIndicator.kt
package com.cherrytree.skinnyyoga.view

import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.widget.TextView

import com.hansoolabs.and.R

import java.util.Timer
import java.util.TimerTask

/**
 * Created by vulpes on 2017. 6. 12..
 */

open class IndexIndicator(private val frame: View, private val adapter: Adapter)
    : IndexScrollerView.IndexIndicator {

    private val parent: ViewGroup
    private val indexView: TextView
    private val totalView: TextView
    private val handler: Handler

    @Volatile
    private var lastTouchedTime: Long = 0
    @Volatile
    private var timer: Timer? = null

    private var animation: Animation? = null

    override val isShown: Boolean
        get() = frame.isShown

    override val isAvailable: Boolean
        get() = adapter.isVisible && adapter.totalSize > 0

    interface Adapter {
        val totalSize: Int
        val isVisible: Boolean
        fun getCurrentIndex(adapterPosition: Int, totalAdapterSize: Int): Int
    }

    init {
        this.parent = frame.parent as ViewGroup
        this.indexView = frame.findViewById<View>(R.id.index) as TextView
        this.totalView = frame.findViewById<View>(R.id.total) as TextView
        handler = Handler(Looper.getMainLooper())
    }

    override fun moveToPosition(scrollProgress: Float) {
        val height = parent.height - frame.height
        val y = scrollProgress * height
        frame.y = y
        lastTouchedTime = System.currentTimeMillis()
    }

    private fun createAnimation(from: Float, to: Float): AlphaAnimation {
        val anim = AlphaAnimation(from, to)
        anim.duration = FADE_ANIMATION_DURATION
        return anim
    }

    override fun show() {
        lastTouchedTime = System.currentTimeMillis()
        startTimer()
        if (frame.visibility != View.VISIBLE) {
            if (animation != null) {
                animation!!.cancel()
            }
            animation = createAnimation(frame.alpha, 1.0f)
            animation!!.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {

                }

                override fun onAnimationEnd(animation: Animation) {
                    frame.alpha = 1.0f
                    this@IndexIndicator.animation = null
                }

                override fun onAnimationRepeat(animation: Animation) {

                }
            })
            frame.visibility = View.VISIBLE
            frame.startAnimation(animation)
        }
    }

    override fun hide() {
        stopTimerIfExist()
        if (frame.visibility != View.GONE) {
            if (animation != null) {
                animation!!.cancel()
            }
            frame.visibility = View.VISIBLE
            animation = createAnimation(frame.alpha, 0.0f)
            animation!!.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {

                }

                override fun onAnimationEnd(animation: Animation) {
                    frame.visibility = View.GONE
                    this@IndexIndicator.animation = null
                }

                override fun onAnimationRepeat(animation: Animation) {

                }
            })
            frame.startAnimation(animation)
        }
    }

    override fun updateIndex(current: Int, total: Int) {
        val index = adapter.getCurrentIndex(current, total)
        val size = adapter.totalSize
        indexView.text = index.toString()
        totalView.text = size.toString()
    }

    @Synchronized
    private fun startTimer() {
        stopTimerIfExist()
        timer = Timer()
        timer!!.schedule(object : TimerTask() {
            override fun run() {
                if (lastTouchedTime + VISIBLE_TIMEOUT < System.currentTimeMillis()) {
                    stopTimerIfExist()
                    handler.post { hide() }
                }
            }
        }, 0, 100)
    }

    @Synchronized
    private fun stopTimerIfExist() {
        if (timer != null) {
            timer!!.cancel()
            timer = null
        }
    }

    companion object {

        private val FADE_ANIMATION_DURATION: Long = 150
        private val VISIBLE_TIMEOUT: Long = 400
    }
}
